Clean up import-expectations.json

gsnedders opened the pull request:
#### 4d09238f938d67cf7d3bf9fa83bac1f1f96e5551
<pre>
Clean up import-expectations.json
<a href="https://bugs.webkit.org/show_bug.cgi?id=232787">https://bugs.webkit.org/show_bug.cgi?id=232787</a>

Reviewed by Jonathan Bedard.

This adds a style checker for import-expectations.json, and fixes
everything in the file to pass the new checker. The checker checks:

 * Schema,
 * That &apos;skip&apos;/&apos;import&apos; status matches the tree,
 * That there are no redundant lines (as values are inherited),
 * That all paths imported are directories.

Note this takes what we have imported as the ground truth (i.e., if
import-expectations.json currently has &apos;skip&apos; but we actually have it
imported, change that to &apos;import&apos;).

Additionally, stop the TestDownloader from adding more redundant lines.

And finally, remove skip lines that refer to paths that no longer exist
in upstream WPT.

* LayoutTests/imported/w3c/resources/import-expectations.json:
* Tools/Scripts/webkitpy/style/checker.py:
(CheckerDispatcher._create_checker):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONChecker.__init__):
(JSONCSSPropertiesChecker.check_codegen_properties):
(JSONImportExpectationsChecker):
(JSONImportExpectationsChecker.check):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker_unittest.py:
* Tools/Scripts/webkitpy/w3c/test_downloader.py:
(TestDownloader.update_import_expectations):

Canonical link: <a href="https://commits.webkit.org/269935@main">https://commits.webkit.org/269935@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34d17667ccf0dae1d64e1b20d8a360ffca0303cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23812 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24865 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25914 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21919 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3462 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22453 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1417 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26508 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/23945 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1194 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21453 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27710 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21729 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25489 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1157 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18850 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1154 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1569 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3067 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
```diff
@@ -418,3 +420,108 @@ def check_codegen_properties(self, property_name, codegen_properties):
+                        "" if exists else "not ",
+                        "" if is_skipped else "not ",
+                    ),
+                )
```
I think this is actually wrong, given:

```json
{
    "web-platform-tests/a": "skip",
    "web-platform-tests/a/b": "import"
}
```

…should have `a` existing despite being skipped. I guess we need to check if we have any expectations where the current path is a prefix.

JonWBedard commented on Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
```diff
@@ -418,3 +423,107 @@ def check_codegen_properties(self, property_name, codegen_properties):
+                "web-platform-tests/"
+            ):
+                self._handle_style_error(
+                    0,
```
Not clear that '0' is the right choice here. This should really be the line number of the error. You might consider searching for `'"{}"'.format(key)` in the original file to assign a line, somewhat like we do in `line_number_from_json_exception`

gsnedders replied:
`line_number_from_json_exception` is looking at the `json.JSONDecodeError` (or `ValueError`, prior to the former's existence), rather than searching the whole file. And really when `json.JSONDecodeError` exists it should just read from that object rather than parsing the message.

`JSONFeaturesChecker` and `JSONCSSPropertiesChecker` both use zero for everything, specifically because we can't accurately attribute errors.

I guess we can (fallibly) guess? But e.g. if someone does something like:

```json
{
    "foo/bar": "web-platform-tests/a/b/c",
    "web-platform-tests/a/b": "import"
    "web-pl\u0061tform-tests/a/b/c": "import"
}
```

…we'll almost certainly blame the wrong line.

gsnedders replied:
Well, I guess we can find all strings in JSON easily enough and parse them and attribute them to lines. It's kinda messy, but I guess it works in general.

JonWBedard reviewed (commented):
Looking at the file right now, it does seem like assigning actual numbers to each of these errors would be pretty easy. I think it's worth doing because if it's not down now, it likely never will be.

gsnedders committed:
Clean up import-expectations.json
https://bugs.webkit.org/show_bug.cgi?id=232787

Reviewed by Jonathan Bedard.

This adds a style checker for import-expectations.json, and fixes
everything in the file to pass the new checker. The checker checks:

 * Schema,
 * That 'skip'/'import' status matches the tree,
 * That there are no redundant lines (as values are inherited),
 * That all paths imported are directories.

Note this takes what we have imported as the ground truth (i.e., if
import-expectations.json currently has 'skip' but we actually have it
imported, change that to 'import').

Additionally, stop the TestDownloader from adding more redundant lines.

And finally, remove skip lines that refer to paths that no longer exist
in upstream WPT.

* LayoutTests/imported/w3c/resources/import-expectations.json:
* Tools/Scripts/webkitpy/style/checker.py:
(CheckerDispatcher._create_checker):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONChecker.__init__):
(JSONCSSPropertiesChecker.check_codegen_properties):
(JSONImportExpectationsChecker):
(JSONImportExpectationsChecker.check):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker_unittest.py:
* Tools/Scripts/webkitpy/w3c/test_downloader.py:
(TestDownloader.update_import_expectations):

Canonical link: https://commits.webkit.org/269935@main
